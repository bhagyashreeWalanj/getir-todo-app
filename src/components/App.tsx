import * as React from 'react'
import { Provider } from 'react-redux'
import ToDoAppContainer from './../containers/ToDoApp.container'
import ToDoListContainer from './../containers/ToDoList.container'

import store from './../store/store'

const App: React.FunctionComponent<{}> = () => {
  return (
    <Provider store={store}>
      <div>
        <ToDoAppContainer />
        <ToDoListContainer />
      </div>
    </Provider>
  )
}

export default App
