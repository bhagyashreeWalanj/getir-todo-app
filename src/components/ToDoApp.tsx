import React, { useState, useEffect } from 'react'
import styles from '../styles/toDoApp.module.scss'
import Alert from './Alert'
import { ITask } from '../models/ITask'

export interface AddToDoProps {
  tasks: ITask[]
  createToDo(title: string, tasks: ITask[]): void
  fetchTasks(): void
  clearTasks(): void
  error: string
}

const ToDoApp = ({
  tasks,
  createToDo,
  fetchTasks,
  clearTasks,
  error,
}: AddToDoProps) => {
  const [alert, setAlert] = useState({ show: false, msg: '', type: '' })
  const [title, setTitle] = useState('')

  const showAlert = (show = false, type = '', msg = '') => {
    setAlert({ show, type, msg })
  }

  // Method to add new task
  const handleCreateToDo = (title: string) => {
    if (title !== '') {
      createToDo(title, tasks)
      if (error !== null) {
        showAlert(true, 'danger', error)
        return
      } else {
        showAlert(true, 'success', 'item added to the list')
        setTitle('')
      }
    } else {
      showAlert(true, 'danger', 'Please add task to save')
      return
    }
  }

  // Method to empty task list
  const handleClearAllTasks = () => {
    clearTasks()
    showAlert(true, 'success', 'List Empty SuccessFully !!')
  }

  useEffect(() => {
    fetchTasks()
  }, [fetchTasks])

  return (
    <div>
      <div className={styles.container}>
        {alert.show && <Alert {...alert} removeAlert={showAlert} />}
        <div className={styles.content}>
          <header className={styles.header}>TO DO LIST</header>
          <div className={styles.inputDivContainer}>
            <input
              className={styles.inputContainer}
              placeholder="Title"
              value={title}
              id={'addToDoInput'}
              onChange={(event) => setTitle(event.target.value)}
            />

            <button
              id={'addButton'}
              type="submit"
              onClick={() => handleCreateToDo(title)}
              className={styles.addButton}
            >
              TO DO
            </button>
            <button
              type="submit"
              onClick={() => handleClearAllTasks()}
              className={styles.clearAllButton}
            >
              CLEAR ALL
            </button>
          </div>
        </div>
      </div>
    </div>
  )
}

export default ToDoApp
