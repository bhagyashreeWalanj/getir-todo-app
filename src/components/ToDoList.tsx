import React, { useState } from 'react'
import styles from '../styles/toDoList.module.scss'
import { default as classnames } from 'classnames'
import { ITask } from '../models/ITask'
import { AiFillDelete } from 'react-icons/ai'
import { FiEdit } from 'react-icons/fi'
import ToDoEditModal from './ToDoEditModal'

export interface IToDoList {
  tasks: ITask[]
  removeTask: (id: string, tasks: ITask[]) => void
  editTask: (id: string, title: string, tasks: ITask[]) => void
}

export function ToDoList({ tasks, removeTask, editTask }: IToDoList) {
  const [taskList, setTaskList] = useState(tasks)
  const [isEditing, setIsEditing] = useState(false)
  const [editID, setEditID] = useState('')
  const [name, setName] = useState('')
  const [isNewTaskModalOpen, setIsNewTaskModalOpen] = useState(false)

  function handleCloseNewTaskModal() {
    setIsNewTaskModalOpen(false)
  }

  const todos = tasks.filter((todo) => !todo.completed)
  const completedTodos = tasks.filter((todo) => todo.completed)

  const handleRemoveTask = (id: string) => {
    removeTask(id, tasks)
  }

  const handleEditTask = (id: string, title: string) => {
    editTask(id, title, tasks)
  }

  const handleCheckBox = (id: string) => {
    tasks = [...tasks].map((task) => {
      if (task.id === id) {
        task.completed = !task.completed
      }
      return task
    })
    setTaskList(tasks)
  }

  const handleEditBtn = (id: string) => {
    const specificItem = tasks.find((item) => item.id === id)
    const title = specificItem?.title
    setIsNewTaskModalOpen(true)
    setName(title!)
    setIsEditing(true)
    setEditID(id)
  }

  return (
    <div className={styles.container}>
      {!tasks ? (
        <div className={styles.noDataDiv}>No Data!</div>
      ) : (
        <div>
          <table className={styles.table}>
            <thead>
              <tr>
                <th className={styles.thead}>Status</th>
                <th className={styles.thead}>Title</th>
                <th className={styles.thead}>Created At</th>
                <th className={styles.thead}>Task Status</th>
                <th className={styles.thead}>Options</th>
              </tr>
            </thead>

            <tbody>
              <tr>
                <td className={styles.taskHeader}>Pending Tasks</td>
              </tr>
              {todos.length === 0 ? (
                <tr className={styles.noDataDiv}>
                  <td>No Pending Tasks!</td>
                </tr>
              ) : (
                todos.map((pendingTask, index) => (
                  <tr
                    key={index}
                    className={'taskTR_pending'}
                    id={`taskTR_pending_${pendingTask.id}`}
                  >
                    <td>
                      <input
                        className={styles.ToDo__checkbox}
                        type="checkbox"
                        id={`checkbox_${pendingTask.id}`}
                        name={`input_${pendingTask.id}`}
                        value={pendingTask.title}
                        checked={pendingTask.completed}
                        onChange={() => handleCheckBox(pendingTask.id)}
                      />
                    </td>

                    <td
                      className={classnames({
                        [styles.strike]: pendingTask.completed,
                      })}
                    >
                      {pendingTask.title}
                    </td>

                    <td
                      className={classnames({
                        [styles.strike]: pendingTask.completed,
                      })}
                    >
                      {pendingTask.createdAt}
                    </td>
                    <td
                      className={classnames({
                        [styles.strike]: pendingTask.completed,
                      })}
                    >
                      {pendingTask.completed ? 'COMPLETE' : 'PENDING'}
                    </td>

                    <td>
                      <div className={styles.options}>
                        <FiEdit
                          id={`editBtn_${pendingTask.id}`}
                          className={styles.ToDo__update}
                          onClick={() => handleEditBtn(pendingTask.id)}
                        >
                          <span className={styles.tooltiptext}>Edit</span>
                        </FiEdit>

                        <AiFillDelete
                          id={`removeBtn_${pendingTask.id}`}
                          className={styles.ToDo__delete}
                          onClick={() => handleRemoveTask(pendingTask.id)}
                        />
                      </div>
                    </td>
                  </tr>
                ))
              )}

              <tr>
                <td className={styles.taskHeader}>Completed Tasks</td>
              </tr>

              {completedTodos.length === 0 ? (
                <tr className={styles.noDataDiv}>
                  <td>-</td>
                </tr>
              ) : (
                completedTodos.map((completedTask, index) => (
                  <tr
                    key={index}
                    className={'taskTR_completed'}
                    id={`taskTR_completed_${completedTask.id}`}
                  >
                    <td>
                      <input
                        className={styles.ToDo__checkbox_completed}
                        type="checkbox"
                        id={`checkbox_${completedTask.id}`}
                        name={`input_${completedTask.id}`}
                        value={completedTask.title}
                        checked={completedTask.completed}
                        onChange={() => handleCheckBox(completedTask.id)}
                      />
                    </td>

                    <td
                      className={classnames({
                        [styles.strike]: completedTask.completed,
                      })}
                    >
                      {completedTask.title}
                    </td>

                    <td
                      className={classnames({
                        [styles.strike]: completedTask.completed,
                      })}
                    >
                      {completedTask.createdAt}
                    </td>
                    <td className={styles.taskComplete}>
                      {completedTask.completed ? 'COMPLETED' : 'PENDING'}
                    </td>
                    <td>
                      <div className={styles.options}>
                        <FiEdit
                          id={`editBtn_${completedTask.id}`}
                          className={styles.ToDo__update}
                          onClick={() => handleEditBtn(completedTask.id)}
                        ></FiEdit>

                        <AiFillDelete
                          id={`removeBtn_${completedTask.id}`}
                          className={styles.ToDo__delete}
                          onClick={() => handleRemoveTask(completedTask.id)}
                        />
                      </div>
                    </td>
                  </tr>
                ))
              )}
            </tbody>
          </table>
        </div>
      )}
      {isEditing ? (
        <ToDoEditModal
          isOpen={isNewTaskModalOpen}
          EditTaskId={editID}
          taskName={name}
          handleEditTask={handleEditTask}
          onRequestClose={handleCloseNewTaskModal}
        />
      ) : (
        ''
      )}
    </div>
  )
}

export default ToDoList
