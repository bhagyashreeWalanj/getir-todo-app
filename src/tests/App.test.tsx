import React from 'react'

import App from '../components/App'
import ToDoApp from '../containers/ToDoApp.container'
import { shallow } from 'enzyme'
import ToDoList from '../containers/ToDoList.container'

describe('render all components', () => {
  const app = shallow(<App />)

  it('should match the snapshot', () => {
    expect(app).toMatchSnapshot()
  })
  it('App Component contains To Do component', () => {
    expect(app.contains(<ToDoApp />)).toBe(true)
  })
  it('App Component contains ToDo Task List component', () => {
    expect(app.contains(<ToDoList />)).toBe(true)
  })
})
