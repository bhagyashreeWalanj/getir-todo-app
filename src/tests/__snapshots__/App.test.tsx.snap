// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render all components should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <div>
    <Connect(ToDoApp) />
    <Connect(ToDoList) />
  </div>
</Provider>
`;
